src/preact/component/iframe.js

```diff
@@ -0,0 +1,159 @@
+import {Loading} from '../../../src/core/loading-instructions';
+import {ReadyState} from '../../../src/ready-state';
+import {dict} from '../../../src/utils/object';
+import {forwardRef} from '../../../src/preact/compat';
+import {useAmpContext, useLoading} from '../../../src/preact/context';
+import {
+  useCallback,
+  useEffect,
+  useImperativeHandle,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from '../../../src/preact';
+
+const NO_HEIGHT_STYLE = dict();
+
+/**
+ * @param {!IframeDef.Props} props
+ * @param {{current: (!IframeDef.Api|null)}} ref
+ * @return {PreactDef.Renderable}
+ */
+export function IframeEmbedWithRef(
+  {
+    allow,
+    allowFullscreen,
+    allowtransparency,
+    name,
+    title,
+    manageMessageHandler,
+    ready = true,
+    requestResize,
+    loading: loadingProp,
+    onReadyState,
+    sandbox,
+    src,
+    ...rest
+  },
+  ref
+) {
+  const {playable} = useAmpContext();
+  const loading = useLoading(loadingProp);
+  const mount = loading !== Loading.UNLOAD;
+
+  const loadedRef = useRef(false);
+  // The `onReadyStateRef` is passed via a ref to avoid the changed values
+  // of `onReadyState` re-triggering the side effects.
+  const onReadyStateRef = useValueRef(onReadyState);
+  const setLoaded = useCallback(
+    (value) => {
+      if (value !== loadedRef.current) {
+        loadedRef.current = value;
+        const onReadyState = onReadyStateRef.current;
+        onReadyState?.(value ? ReadyState.COMPLETE : ReadyState.LOADING);
+      }
+    },
+    [onReadyStateRef]
+  );
+
+  const iframeRef = useRef(null);
+  const [heightStyle, setHeightStyle] = useState(NO_HEIGHT_STYLE);
+  const [opacity, setOpacity] = useState(0);
+
+  // Component API: IframeEmbedDef.Api.
+  useImperativeHandle(
+    ref,
+    () => ({
+      // Standard Bento
+      get readyState() {
+        return loadedRef.current ? ReadyState.COMPLETE : ReadyState.LOADING;
+      },
+    }),
+    []
+  );
+
+  // Reset readyState to "loading" when an iframe is unloaded. Has to be
+  // a `useLayoutEffect` to avoid race condition with a future "load" event.
+  // A race condition can happen when a `useEffect` would be executed
+  // after a future "load" is dispatched.
+  useLayoutEffect(() => {
+    if (!mount) {
+      setLoaded(false);
+    }
+  }, [mount, setLoaded]);
+
+  // Pause if the post goes into a "paused" context.
+  useEffect(() => {
+    const iframe = iframeRef.current;
+    if (!playable && iframe) {
+      // Resetting the `src` will reset the iframe and pause it. It will force
+      // the reload of the whole iframe. But it's the only reliable option
+      // to force pause.
+      iframe.src = iframe.src;
+    }
+  }, [playable]);
+
+  useLayoutEffect(() => {
+    if (!iframeRef.current || !mount) {
+      return;
+    }
+    return manageMessageHandler(iframeRef, (heightOnSuccess) => {
+      if (requestResize) {
+        requestResize(heightOnSuccess);
```

Comment:
I think `requestResize` should be called by the actual component from within the `props.messageHandler()`. Otherwise the resize format is always very different from an iframe to iframe.